scalding-core/src/main/scala/com/twitter/scalding/serialization/JavaStreamEnrichments.scala

```diff
@@ -0,0 +1,221 @@
+  }
+
+  /**
+   * This has a lot of methods from DataInputStream without
+   * having to allocate to get them
+   * This code is similar to those algorithms
+   */
+  implicit class RichInputStream(val s: InputStream) extends AnyVal {
+    /**
+     * If s supports marking, we mark it. Otherwise we read the needed
+     * bytes out into a ByteArrayStream and return that.
+     * This is intended for the case where you need possibly
+     * read size bytes but may stop early, then skip this exact
+     * number of bytes.
+     * Intended use is:
+     * {code}
+     * val size = 100
+     * val marked = s.markOrBuffer(size)
+     * val y = fn(marked)
+     * marked.reset
+     * marked.skipFully(size)
+     * {/code}
+     */
+    def markOrBuffer(size: Int): InputStream = {
+      val ms = if (s.markSupported) s else {
+        val buf = new Array[Byte](size)
+        s.readFully(buf)
+        new ByteArrayInputStream(buf)
+      }
+      // Make sure we can reset after we read this many bytes
+      ms.mark(size)
+      ms
+    }
+
+    def readBoolean: Boolean = (readUnsignedByte != 0)
+
+    /**
+     * Like read, but throws eof on error
+     */
+    def readByte: Byte = readUnsignedByte.toByte
+
+    def readUnsignedByte: Int = {
+      val c1 = s.read
+      if (c1 < 0) eof else c1
+    }
+    def readUnsignedShort: Int = {
+      val c1 = s.read
+      val c2 = s.read
+      if ((c1 | c2) < 0) eof else ((c1 << 8) | c2)
+    }
+
+    final def readFully(bytes: Array[Byte]): Unit = readFully(bytes, 0, bytes.length)
+
+    final def readFully(bytes: Array[Byte], offset: Int, len: Int): Unit = {
+      if (len < 0) throw new IndexOutOfBoundsException()
+
+      @annotation.tailrec
+      def go(o: Int, l: Int): Unit =
+        if (l == 0) ()
+        else {
+          val count = s.read(bytes, o, l)
+          if (count < 0) eof
+          else go(o + count, l - count)
+        }
+      go(offset, len)
+    }
+
+    def readDouble: Double = java.lang.Double.longBitsToDouble(readLong)
+    def readFloat: Float = java.lang.Float.intBitsToFloat(readInt)
+
+    def readInt: Int = {
+      val c1 = s.read
+      val c2 = s.read
+      val c3 = s.read
+      val c4 = s.read
+      // This is the algorithm from DataInputStream
+      if ((c1 | c2 | c3 | c4) < 0) eof else ((c1 << 24) | (c2 << 16) | (c3 << 8) | c4)
+    }
+    def readLong: Long = {
+      // This is the algorithm from DataInputStream
+      val buf = new Array[Byte](8)
+      readFully(buf)
+      (buf(0).toLong << 56) +
+        ((buf(1) & 255).toLong << 48) +
+        ((buf(2) & 255).toLong << 40) +
+        ((buf(3) & 255).toLong << 32) +
+        ((buf(4) & 255).toLong << 24) +
+        ((buf(5) & 255) << 16) +
+        ((buf(6) & 255) << 8) +
+        (buf(7) & 255)
+    }
+
+    def readShort: Short = {
+      val c1 = s.read
+      val c2 = s.read
+      // This is the algorithm from DataInputStream
+      if ((c1 | c2) < 0) eof else ((c1 << 8) | c2).toShort
```

Comment:
There is an unsignedShort one (that returns an Int, because Java). This one, will interpret the bytes as a signed number. Java does this for the low 16 bits of an Int

```scala
scala> 65000.toShort
res21: Short = -536
```